tests/pagination/src/commonTest/kotlin/OffsetBasedWithArrayPaginationTest.kt

```diff
@@ -0,0 +1,194 @@
+    ))
+    client.apolloStore.writeOperation(query4, data4)
+    dataFromStore = client.apolloStore.readOperation(query1)
+    expectedData = UsersOffsetBasedWithArrayQuery.Data(listOf(
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("40", "Paul", "paul@a.com", "User"),
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("41", "Mary", "mary@a.com", "User"),
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("42", "John", "john@a.com", "User"),
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("43", "Jane", "jane@a.com", "User"),
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("44", "Peter", "peter@a.com", "User"),
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("45", "Alice", "alice@a.com", "User"),
+        UsersOffsetBasedWithArrayQuery.UsersOffsetBasedWithArray("46", "Bob", "bob@a.com", "User"),
+    ))
+    assertEquals(expectedData, dataFromStore)
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private class OffsetPaginationMetadataGenerator(private val fieldName: String) : MetadataGenerator {
+    override fun metadataForObject(obj: Any?, context: MetadataGeneratorContext): Map<String, Any?> {
+      if (context.field.name == fieldName) {
+        return mapOf(
+            "offset" to context.field.resolveArgument("offset", context.variables) as Int,
+        )
+      }
+      return emptyMap()
+    }
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private class OffsetPaginationRecordMerger : RecordMerger {
+    override fun merge(existing: Record, incoming: Record, newDate: Long?): Pair<Record, Set<String>> {
+      val changedKeys = mutableSetOf<String>()
+      val mergedFields = existing.fields.toMutableMap()
+      val mergedMetadata = existing.metadata.toMutableMap()
+      val date = existing.date?.toMutableMap() ?: mutableMapOf()
+
+      for ((fieldKey, incomingFieldValue) in incoming.fields) {
+        val hasExistingFieldValue = existing.fields.containsKey(fieldKey)
+        val existingFieldValue = existing.fields[fieldKey]
+        if (!hasExistingFieldValue || existingFieldValue != incomingFieldValue) {
+          val existingOffset = existing.metadata[fieldKey]!!["offset"] as? Int
+          val incomingOffset = incoming.arguments[fieldKey]!!["offset"] as? Int
+          if (existingOffset == null || incomingOffset == null) {
+            mergedFields[fieldKey] = incomingFieldValue
+            mergedMetadata[fieldKey] = incoming.metadata[fieldKey] as Map<String, Any?>
+          } else {
+            val existingList = existing[fieldKey] as List<*>
+            val incomingList = incomingFieldValue as List<*>
+            val mergedList = mergeLists(existingList, incomingList, existingOffset, incomingOffset)
+            mergedFields[fieldKey] = mergedList
+            mergedMetadata[fieldKey] = mapOf("offset" to min(existingOffset, incomingOffset))
```

Comment:
Using `min` doesn't take into account the case where existing and incoming do not join (the early returns in `mergeLists`)